syntax = "proto2";

option java_package = "org.simgrid.rsg";

package simgrid.rsg;

enum Type {
  //Actor
  CMD_SLEEP = 1;
  CMD_EXEC = 2;
  CMD_SEND = 3;
  CMD_RECV = 4;

  //Engine
  CMD_ENGINE_GETCLOCK = 50;

  //Mailbox
  CMD_MB_CREATE = 5;

  //Remote-Simgrid specific
  CMD_QUIT = 90;
}

message Request { 
  // The kind of request that we have today
  required Type type = 99;

  //
  //// Actor
  //

  // Sleep request
  message SleepRequest { required double duration = 1; }
  optional SleepRequest sleep = 1;
  
  // Exec request
  message ExecRequest { required double flops = 1; }
  optional ExecRequest exec = 2;

  // Send request
  message SendRequest { 
    required uint64 mbox = 1;
    required bytes content = 2;
    required uint64 contentSize = 3;
    optional uint64 simulatedSize = 4 [default = 0];
  }
  optional SendRequest send = 3;

  // Recv request
  message RecvRequest { required uint64 mbox = 1; }
  optional RecvRequest recv = 4;

  //
  //// Engine
  //

  // CMD_ENGINE_GETCLOCK request: no args


  //
  //// Mailbox
  //

  // Mailbox Create
  message MBCreateRequest { required string name = 1; }
  optional MBCreateRequest mbCreate = 5;


  //
  //// Remote-Simgrid specific
  //

  // QUIT request: no args

}

message Answer {
  // The kind of request that we are replying to today
  required Type type = 99;
  required double clock = 100;


  //
  //// Actor
  //

  // Sleep request: no returned type
  // Exec request: no returned type
  // Send request: no returned type
  message RecvAnswer { 
  	required bytes content = 1;
  	required uint64 contentSize = 2;
  }
  optional RecvAnswer recv = 4;


  //
  //// Engine
  //

  // CMD_ENGINE_GETCLOCK request: returned info in header


  //
  //// Mailbox
  //

  message MBCreateAnswer { required uint64 remoteAddr = 1; }
  optional MBCreateAnswer mbCreate = 5;


  //
  //// Remote-Simgrid specific
  //

  // QUIT request: no returned type

}
